modeshape-jcr/src/main/java/org/modeshape/jcr/value/binary/DatabaseBinaryStore.java

```diff
@@ -173,14 +174,29 @@ public BinaryValue storeValue( InputStream stream, final boolean markAsUnused )
                         database.restoreContent(connection, Collections.singletonList(key));
                     }
                 } else {
-                    // store the content
-                    database.insertContent(key, temp.getStream(), temp.getSize(), connection);
-                    if (markAsUnused) {
-                        database.markUnused(Collections.singletonList(key), connection);
+                    try (InputStream is = temp.getStream()) {
+                        // store the content
+                        database.insertContent(key, is, temp.getSize(), connection);
+                        if (markAsUnused) {
+                            database.markUnused(Collections.singletonList(key), connection);
+                        }
                     }
                 }
                 return new StoredBinaryValue(DatabaseBinaryStore.this, key, temp.getSize());
             });
+        } catch (BinaryStoreException e) {
+            if (e.getCause() instanceof SQLException) {
+                // under certain conditions - e.g. in a cluster - someone else may have already inserted the binary
+                // so try reading again
+                 return dbCall(connection -> {
+                    if (database.contentExists(key, !markAsUnused, connection)) {
+                        return new StoredBinaryValue(DatabaseBinaryStore.this, key, temp.getSize());        
+                    }
+                    // nothing there, so rethrow the original exception
+                    throw e;
+                });
+            }
+            throw e;
```

Comment:
@rhauch if you prefer I can look at either of the above 2 options, although I'm not 100% convinced the changes would be significant